File: concepts/07_modules_packages_and_loading/03_initialization_caching_cycles_and_dynamic_loading/julia/test_02_failed_import_and_cache_state.jl
# polyglot-family: modules_packages_and_loading
# polyglot-concept: initialization_caching_cycles_and_dynamic_loading
# polyglot-related: languages/julia/tooling_and_runtime/14_pkg_environments_and_code_loading/
# polyglot-related+: test_109_find_package_and_source_identity.jl
#
# 共同问题：加载失败是否污染成功缓存；后续解析能否区分未找到与已加载。
# 对照观察：不存在的 PkgId 使 require 失败且不会出现在 loaded_modules；find_package 仍返回 nothing。

using Test
using UUIDs

@testset "失败 require 不建立成功缓存项" begin
    package_id = Base.PkgId(UUID("d44df542-1aea-4fd6-80c0-f2c50d2a56ff"), "MissingPolyglotPackage")
    @test_throws ArgumentError Base.require(package_id)
    @test_throws ArgumentError Base.require(package_id)
    @test Base.find_package(package_id.name) === nothing
end

File: concepts/07_modules_packages_and_loading/03_initialization_caching_cycles_and_dynamic_loading/rust/test_02_failed_import_and_cache_state.rs
// polyglot-family: modules_packages_and_loading
// polyglot-concept: initialization_caching_cycles_and_dynamic_loading
// polyglot-related: languages/rust/tests/tooling_and_runtime/
// polyglot-related+: 10_modules_visibility_linkage_and_initialization/test_080_dynamic_loading_and_reload_boundary.rs
//
// 共同问题：加载失败是否污染缓存；能否运行时重试、重新加载或动态选择模块。
// 对照观察：missing module 是 build error，不产生 runtime cache entry；std 没有通用 import retry/reload/dlopen API。

use polyglot_rust_harness::assert_compile_fails;

#[test]
fn comparison() {
    assert_compile_fails(
        "mod missing; fn main() {}",
        &["file not found for module `missing`"],
    );
    assert_compile_fails(
        "fn main() { std::module::reload(\"plugin\"); }",
        &["could not find", "std"],
    );
    assert_compile_fails(
        "fn main() { std::dlopen::open(\"plugin.so\"); }",
        &["could not find", "std"],
    );
}
